### PR TITLE
Test the correctness of optimizer_wrapper for momentum SGD

### PR DESCRIPTION
**Pull request by mhaoli**
Part of #1123, test the correctness of optimizer_wrapper for momentum SGD.

### CODE REVIEW

File: elasticdl/python/tests/embedding_test_module.py [hunk 1]
```diff
@@ -4,23 +4,65 @@
 from elasticdl.python.elasticdl.layers.embedding import Embedding
 
 
-class CustomModel(tf.keras.Model):
-    def __init__(self, output_dim=16):
-        super(CustomModel, self).__init__(name="embedding_test_model")
+class EdlEmbeddingModel(tf.keras.Model):
+    def __init__(self, output_dim=16, weights=None):
```
Comment (skydoorkai) [inline elasticdl/python/tests/embedding_test_module.py:8]: Add comments stating that weights are the initialization values if set.

File: elasticdl/python/tests/optimizer_wrapper_test.py [hunk 2]
```diff
@@ -14,6 +21,139 @@
...
+    random_seed,
+):
+    """
+    Generate data for training. `is_sparse=False` means we require that all
+    embedding ids appear in every batch data.
+    """
+    random.seed(random_seed)
+    X, Y = [], []
+    if not is_sparse:
+        assert input_length > input_dim, "`input_length` should be large "
```
Comment (skydoorkai) [inline elasticdl/python/tests/optimizer_wrapper_test.py]: large -> larger

File: elasticdl/python/tests/optimizer_wrapper_test.py [hunk 3]
```diff
@@ -333,6 +473,67 @@ def test_report_to_kv_store(self):
...
+                w2 = np.concatenate(mock_kv_store.lookup(keys)[0]).reshape(
+                    4, -1
+                )
+                self.assertTrue((w1 - w2 < 0.0001).all())
+            else:
+                for w1, w2 in zip(layer1.weights, layer2.weights):
+                    self.assertTrue((w1 - w2 < 0.0001).numpy().all())
+
+    def test_correctness(self):
+        self._test_correctness(SGD, 0.1, 0.5)
```
Comment (skydoorkai) [inline elasticdl/python/tests/optimizer_wrapper_test.py]: Use `kwargs` to pass arguments so it is clear what these arguments are.

File: elasticdl/python/tests/optimizer_wrapper_test.py (hunk 2 above)
Comment (mhaoli) [reply elasticdl/python/tests/optimizer_wrapper_test.py]: Fixed.

File: elasticdl/python/tests/optimizer_wrapper_test.py (hunk 3 above)
Comment (mhaoli) [reply elasticdl/python/tests/optimizer_wrapper_test.py]: Fixed.

File: elasticdl/python/tests/embedding_test_module.py (hunk 1 above)
Comment (mhaoli) [reply elasticdl/python/tests/embedding_test_module.py:8]: Added.

File: elasticdl/python/master/optimizer_wrapper.py [hunk 4]
```diff
@@ -262,6 +262,7 @@ def _generate_lookup_keys(self, grads_and_vars):
             # de-duplicate gradient's indices
             unique_ids, indices = tf.unique(grad.indices)
             unique_ids = unique_ids.numpy()
+            # TODO: support grads_and_vars with duplicated layer name
```
Comment (terrytangyuan) [inline elasticdl/python/master/optimizer_wrapper.py]: Is this going to be fixed soon? If not, log a warning when detected?

File: elasticdl/python/tests/embedding_test_module.py [hunk 5]
```diff
@@ -4,23 +4,83 @@
...
+        """
+        Arguments:
+            output_dim: An Integer. It is the output dimension of embedding
+                layers in `EdlEmbeddingModel`.
+            weights: A numpy ndarray list. Unless `weights` is None, dense
```
Comment (terrytangyuan) [inline elasticdl/python/tests/embedding_test_module.py]: "Unless..." -> "If ... not None"

File: elasticdl/python/tests/embedding_test_module.py [hunk 6]
```diff
@@ -4,23 +4,83 @@
...
+                layer initializes its weights using `weights`.
+        """
+        super(EdlEmbeddingModel, self).__init__(
+            name="test_model_with_edl_embedding"
+        )
         self.output_dim = output_dim
+        if weights:
+            if len(weights) != 2:
+                raise ValueError(
+                    "test_model_with_edl_embedding constructor receives "
```
Comment (terrytangyuan) [inline elasticdl/python/tests/embedding_test_module.py]: test_model_with_edl_embedding -> why not "EdlEmbeddingModel"?

File: elasticdl/python/tests/embedding_test_module.py [hunk 7]
```diff
@@ -4,23 +4,83 @@
...
         self.concat = Concatenate()
-        self.dense = Dense(1, input_shape=(output_dim * 3,))
+        self.dense = Dense(1, weights=weights)
+        self.flatten = Flatten()
+
+    def call(self, inputs, training=False):
+        f1 = self.embedding_1(inputs["f1"])
+        f2 = self.embedding_1(inputs["f2"])
+        f3 = self.embedding_2(inputs["f3"])
+        x = self.concat([f1, f2, f3])
```
Comment (terrytangyuan) [inline elasticdl/python/tests/embedding_test_module.py]: We don't need to assign f1, f2, f3. You can use use them directly in concat.

File: elasticdl/python/tests/optimizer_wrapper_test.py [hunk 8]
```diff
@@ -14,6 +21,143 @@
...
+    input_dim,
+    is_sparse,
+    random_seed,
+):
+    """
+    Generate data for training. `is_sparse=False` means we require that all
+    embedding ids appear in every batch data.
+    """
+    random.seed(random_seed)
+    X, Y = [], []
```
Comment (terrytangyuan) [inline elasticdl/python/tests/optimizer_wrapper_test.py:38]: Maybe use "features" and "labels" instead?

File: elasticdl/python/tests/optimizer_wrapper_test.py [hunk 9]
```diff
@@ -14,6 +21,143 @@
...
+):
+    """Train model with optimizer wrapper."""
+    tf.random.set_seed(random_seed)
+    optimizer = OptimizerWrapper(opt_keras, None, embed_dims)
+
+    # initialization process related to embedding layer and optimizer wrapper
+    embed_layers = []
+    for layer in model.layers:
+        if isinstance(layer, Embedding):
+            embed_layers.append(layer)
```
Comment (terrytangyuan) [inline elasticdl/python/tests/optimizer_wrapper_test.py]: I remember we have a utility function for this? Reuse `model_helper.find_layer`?

File: elasticdl/python/master/optimizer_wrapper.py (hunk 4 above)
Comment (mhaoli) [reply elasticdl/python/master/optimizer_wrapper.py]: Added.

File: elasticdl/python/tests/embedding_test_module.py (hunk 5 above)
Comment (mhaoli) [reply elasticdl/python/tests/embedding_test_module.py]: Fixed

File: elasticdl/python/tests/embedding_test_module.py (hunk 6 above)
Comment (mhaoli) [reply elasticdl/python/tests/embedding_test_module.py]: Use "EdlEmbeddingModel" here.

File: elasticdl/python/tests/embedding_test_module.py (hunk 7 above)
Comment (mhaoli) [reply elasticdl/python/tests/embedding_test_module.py]: Fixed.

File: elasticdl/python/tests/optimizer_wrapper_test.py (hunk 8 above)
Comment (mhaoli) [reply elasticdl/python/tests/optimizer_wrapper_test.py:38]: "features" and "labels" are used for minibatch data in this test, i.e. 

```
for features, labels in zip(X, Y)
```

File: elasticdl/python/tests/optimizer_wrapper_test.py (hunk 9 above)
Comment (mhaoli) [reply elasticdl/python/tests/optimizer_wrapper_test.py]: Fixed.

**Review by skydoorkai — APPROVED** [review]
